src/flexible_type/ndarray.hpp

```diff
@@ -0,0 +1,715 @@
+   */
+  container_type& raw_elements() {
+    return *m_elem;
+  }
+
+  /**
+   * Returns a reference to all the elements in a linear layout; if is_full()
+   * is false, this will include unindexable elements.
+   */
+  const container_type& raw_elements() const {
+    return *m_elem;
+  }
+
+  /**
+   * Returns a reference to all the elements in a linear layout, is_full()
+   * must be true.
+   *
+   * Note that if the memory used by this array is shared, this may have
+   * unintentional side effects (changing other arrays).
+   */
+  container_type& elements() {
+    ensure_unique();
+    ASSERT_TRUE(is_full());
+    return *m_elem;
+  }
+  /**
+   * Returns a const reference to all the elements in a linear layout, is_full()
+   * must be true.
+   */
+  const container_type& elements() const {
+    ASSERT_TRUE(is_full());
+    return *m_elem;
+  }
+
+  /**
+   * Returns a const reference to the shape.
+   */
+  const index_range_type& shape() const {
+    return m_shape;
+  }
+
+  /**
+   * Returns a const reference to the stride.
+   */
+  const index_range_type& stride() const {
+    return m_stride;
+  }
+
+  /**
+   * Returns a const reference to the stride.
+   */
+  index_type start() const {
+    return m_start;
+  }
+
+  /**
+   * Returns the number of elements in the array.
+   *
+   * This is equivalent to the product of the values in the shape array.
+   * Note that this may not be the same as elements().size().
+   */
+  size_t num_elem() const {
+    if (m_shape.size() == 0) return 0;
+    if (m_elem == nullptr) return 0;
+    size_t p = 1;
+    for (size_t s: m_shape) {
+      p = p * s;
+    }
+    return p;
+  }
+
+  /**
+   * Returns true if every element in elements() is reachable by an
+   * N-d index.
+   */
+  bool is_full() const {
+    return m_start == 0 && 
+        num_elem() == m_elem->size() && 
+        last_index() == m_elem->size();
+  }
+
+  /**
+   * Returns true if the shape and stride of the array is laid out 
+   * correctly such at all array indices are within elements().size().
+   * 
+   * An ndarray can be invalid for instance, if the stride is too large,
+   * or if the shape is larger than the total number of elements.
+   */
+  bool is_valid() const {
+    return m_shape.size() == m_stride.size() &&   // shape and stride line up
+          num_elem() + m_start <= m_elem->size() &&  // num_elements (as computed by shape) is in m_elem
+          last_index() + m_start <= m_elem->size();  // max index (as computed by stride( is in m_elem
+  }
+
+  /** 
+   * Returns true if the stride is ordered canonically.
+   * The strides must be non-increasing and non-zero.
+  */
+  bool has_canonical_stride() const {
+    if (m_stride.size() == 0) return true;
+    if (m_stride[0] == 0) return false;
+    for (size_t i = 1; i < m_stride.size(); ++i) {
+      if (m_stride[i] == 0 || m_stride[i - 1] < m_stride[i]) {
+        return false;
+      }
+    }
+    return true;
```

Comment:
Need to test that final stride equals 1?

Reply:
Yeah. This just tests for C-ordering. is_full() will test if it is complete.